lib/atlas/parser.rb

```diff
@@ -58,11 +58,20 @@ def process_child_node
         current_node << node
 
         if current_node.limit_reached?
-          @current_node = current_node.parent
+          set_current_node_to_ancestor
         elsif !node.limit_reached?
           @current_node = node
         end
       end
+
+      def set_current_node_to_ancestor
+        loop do
+          @current_node = current_node.parent
+
+          # If the parent is also full, keep going up the ancestry
+          break unless @current_node.limit_reached?
```

Comment:
This was causing problems because the `NOT(INCLUDES(...))` would exit the `INCLUDES()` node but then the `@current_node` would be set to the `NOT()` node instead of the parent of `NOT()`.

Reply:
Am I missing something here? It seems like `current_node` is the same on every iteration of the `loop`.

Reply:
this happens on every loop:

```
@current_node = current_node.parent
```

It's like a tree data structure, it's going up towards the root.

Reply:
Is the root node always an NaryNode? If that's not true then the root could be full as well. I'm wondering if the loop short circuit code should contemplate such case: 

```
break if !@current_node || !@current_node.limit_reached? 
```

Reply:
Good call, updated: bcdeabb8210c2b421145db0714dd27063e2b8795